src/util/ProjectIO.as

```diff
@@ -207,48 +207,113 @@ public class ProjectIO {
-		// Load all images in all costumes from their image data, then call whenDone.
-		function imageDecoded():void {
-			for each (var o:* in imageDict) {
-				if (o == 'loading...') return; // not yet finished loading
+		// This should be called on success or failure of each image
+		function imageDone():void {
+			if (--numImagesToDecode == 0) {
+				allImagesLoaded(objList, imageDict, whenDone, fail);
 			}
-			allImagesLoaded();
 		}
-		var error:Boolean = false;
-		function decodeError():void {
-			if (error) return;
-			error = true;
-			if (fail != null) fail();
-		}
-		function allImagesLoaded():void {
-			if (error) return;
-			for each (c in allCostumes) {
+
+		var numImagesToDecode:int = 1; // start at 1 to prevent early finish
+		var imageDict:Dictionary = new Dictionary(); // maps image data to BitmapData
+		for each (var obj:ScratchObj in objList) {
+			for each (var c:ScratchCostume in obj.costumes) {
 				if ((c.baseLayerData != null) && (c.baseLayerBitmap == null)) {
-					var img:* = imageDict[c.baseLayerData];
-					if (img is BitmapData) c.baseLayerBitmap = img;
-					if (img is SVGElement) c.setSVGRoot(img, false);
+					++numImagesToDecode;
+					if (ScratchCostume.isSVGData(c.baseLayerData)) {
+						decodeSVG(c.baseLayerData, imageDict, imageDone);
+					}
+					else {
+						decodeImage(c.baseLayerData, imageDict, imageDone, imageDone);
+					}
+				}
+				if ((c.textLayerData != null) && (c.textLayerBitmap == null)) {
+					++numImagesToDecode;
+					decodeImage(c.textLayerData, imageDict, imageDone, imageDone);
 				}
-				if ((c.textLayerData != null) && (c.textLayerBitmap == null)) c.textLayerBitmap = imageDict[c.textLayerData];
 			}
-			for each (c in allCostumes) c.generateOrFindComposite(allCostumes);
-			whenDone();
 		}
+		// decrement the artificial 1 and also handle the case where no images are present.
+		imageDone();
+	}
 
-		var c:ScratchCostume;
-		var allCostumes:Array = [];
-		for each (var o:ScratchObj in objList) {
-			for each (c in o.costumes) allCostumes.push(c);
+	private function allImagesLoaded(objList:Array, imageDict:Dictionary, whenDone:Function, fail:Function):void {
+		var errorCostumes:Vector.<ScratchCostume> = new Vector.<ScratchCostume>(2);
+
+		function makeErrorImage(obj:ScratchObj, c:ScratchCostume):* {
+			// TODO: translation?
+			if (!errorDialog) {
+				errorDialog = new DialogBox();
+				errorDialog.addTitle('Image loading error');
+				errorDialog.addText('At least one backdrop or costume failed to load:\n');
+			}
+			if (obj.isStage) {
+				errorDialog.addText('Backdrop: ' + c.costumeName + '\n');
+			}
+			else {
+				errorDialog.addText('Sprite: ' + obj.objName + '\nCostume: ' + c.costumeName + '\n');
+			}
+
+			var errorCostumeIndex:int = 0 + obj.isStage;
+			var errorCostume:ScratchCostume = errorCostumes[errorCostumeIndex] =
+					errorCostumes[errorCostumeIndex] || ScratchCostume.emptyBitmapCostume('', obj.isStage);
+			return errorCostume.baseLayerBitmap;
```

Comment:
You might argue that we should replace a broken vector costume with a new, empty vector costume. I agree in principle, but at this point in the code we've already cleared out any information we could use to reliably determine whether or not the costume is supposed to be vector.
Note that `ScratchCostume.isSVGData()` only returns `true` if the data is complete and well-formed SVG, so corrupted SVG data looks like bitmap data if you use that method.